File: go/internal/parser/parser.go
// Package parser implements the GridMD document parser (SPEC §2–§10, Appendix
// A). It produces a block tree; semantic checks live in package validate.
package parser

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/lprhodes/grid-md/go/internal/yamlsubset"
)

// ReservedKinds is the closed set of directive kinds in 0.1.
var ReservedKinds = map[string]bool{
	"sheet": true, "grid": true, "spill-cache": true, "table": true, "cf": true,
	"validation": true, "filter": true, "chart": true, "sparklines": true,
	"pivot": true, "slicer": true, "image": true, "shape": true, "textbox": true,
	"checkbox": true, "comments": true, "outline": true, "page": true,
	"query": true, "script": true, "scenario": true, "raw": true,
}

// Diag is a line-tagged diagnostic message.
type Diag struct {
	Line int
	Msg  string
}

// Size is a directive `size WxH` argument in pixels.
type Size struct {
	W, H int
}

// InfoArgs holds a fence info-string's parsed arguments.
type InfoArgs struct {
	Positional []string
	Flags      map[string]string
	Anchor     *string
	Size       *Size
}

// Row is one payload pipe row of a grid/table/spill-cache block.
type Row struct {
	Cells []string
	Line  int
}

// Block is either an @-directive or a fenced directive.
type Block interface{ isBlock() }

// AtBlock is a parsed @ directive.
type AtBlock struct {
	Line       int
	TargetText string
	ScalarText *string
	Props      map[string]interface{}
	Body       map[string]interface{}
}

func (b *AtBlock) isBlock() { _ = b }

// FenceBlock is a parsed fenced directive.
type FenceBlock struct {
	Line    int
	Kind    string
	Args    InfoArgs
	Body    []string
	Rows    []Row
	Meta    interface{}
	Code    string
	Payload string
}

func (b *FenceBlock) isBlock() { _ = b }

// Sheet is a level-1 heading and its blocks.
type Sheet struct {
	Name   string
	Line   int
	Blocks []Block
}

// Document is the full parsed tree.
type Document struct {
	Frontmatter    map[string]interface{}
	WorkbookBlocks []Block
	Sheets         []*Sheet
	Errors         []Diag
	Warnings       []Diag
	Mode           string

	// Stats populated by validate.
	StatsDefs   int
	StatsBlocks int
}

// ParseYaml parses a YAML fragment, appending any diagnostics at the given line
// to errors. An empty fragment yields an empty map.
func ParseYaml(text string, line int, errors *[]Diag) interface{} {
	v, errs := yamlsubset.Decode(text)
	for _, e := range errs {
		*errors = append(*errors, Diag{Line: line, Msg: e})
	}
	return v
}

// TryProps returns the flow-map props for an @ directive when text parses to a
// mapping of identifier keys with non-null values (SPEC Appendix A).
func TryProps(text string) (map[string]interface{}, bool) {
	return yamlsubset.TryProps(text)
}

// FindPropsSplit performs the right-edge props split (SPEC §9.1 / Appendix A):
// the last brace-balanced {…} group (double quotes respected) that runs to
// end-of-line and is preceded by whitespace.
func FindPropsSplit(text string) (scalarText string, propsText string, ok bool) {
	if !strings.HasSuffix(text, "}") {
		return text, "", false
	}
	inQ := false
	depth := 0
	start := -1
	lastStart, lastEnd := -1, -1
	for i := 0; i < len(text); i++ {
		ch := text[i]
		if ch == '"' {
			inQ = !inQ
			continue
		}
		if inQ {
			continue
		}
		if ch == '{' {
			if depth == 0 {
				start = i
			}
			depth++
		} else if ch == '}' {
			depth--
			if depth == 0 && start != -1 {
				lastStart, lastEnd = start, i
			}
			if depth < 0 {
				return text, "", false
			}
		}
	}
	if lastStart == -1 {
		return text, "", false
	}
	s, e := lastStart, lastEnd
	if e != len(text)-1 || s == 0 || text[s-1] != ' ' {
		return text, "", false
	}
	return strings.TrimRight(text[:s], " \t"), text[s:], true
}

// SplitPipeRow splits a pipe row into trimmed cell strings; a backslash escapes
// the next character. Returns ok=false for a malformed row.
func SplitPipeRow(rawLine string) ([]string, bool) {
	line := strings.TrimRight(rawLine, " \t\r\n\v\f")
	if !strings.HasPrefix(line, "|") || len(line) < 2 {
		return nil, false
	}
	var cells []string
	var cell strings.Builder
	opened := false
	for i := 0; i < len(line); i++ {
		ch := line[i]
		if ch == '\\' && i+1 < len(line) {
			cell.WriteByte(line[i+1])
			i++
			continue
		}
		if ch == '|' {
			if !opened {
				opened = true
				continue
			}
			cells = append(cells, strings.TrimSpace(cell.String()))
			cell.Reset()
			continue
		}
		cell.WriteByte(ch)
	}
	if strings.TrimSpace(cell.String()) != "" {
		return nil, false // no unescaped closing pipe
	}
	return cells, true
}

var infoTokenRe = regexp.MustCompile(`"((?:[^"]|"")*)"|\S+`)
var flagRe = regexp.MustCompile(`^([A-Za-z][A-Za-z0-9-]*)=(.*)$`)
var sizeRe = regexp.MustCompile(`^(\d+)x(\d+)$`)
var flagQuoteRe = regexp.MustCompile(`(?s)^"(.*)"$`)

type infoToken struct {
	v string
	q bool
}

// ParseInfoArgs parses a fence info string into positional args, flags, an
// anchor, and a size.
func ParseInfoArgs(rest string, line int, errors *[]Diag) InfoArgs {
	out := InfoArgs{Flags: map[string]string{}}
	var tokens []infoToken
	for _, m := range infoTokenRe.FindAllStringSubmatchIndex(rest, -1) {
		if m[2] != -1 {
			inner := rest[m[2]:m[3]]
			tokens = append(tokens, infoToken{v: strings.ReplaceAll(inner, `""`, `"`), q: true})
		} else {
			tokens = append(tokens, infoToken{v: rest[m[0]:m[1]], q: false})
		}
	}
	for k := 0; k < len(tokens); k++ {
		tok := tokens[k]
		if !tok.q && tok.v == "at" {
			if k+1 >= len(tokens) {
				*errors = append(*errors, Diag{Line: line, Msg: "`at` requires an anchor"})
				break
			}
			k++
			a := tokens[k].v
			out.Anchor = &a
			continue
		}
		if !tok.q && tok.v == "size" {
			if k+1 >= len(tokens) {
				*errors = append(*errors, Diag{Line: line, Msg: "`size` requires WxH (e.g. 480x320)"})
				continue
			}
			k++
			sm := sizeRe.FindStringSubmatch(tokens[k].v)
			if sm == nil {
				*errors = append(*errors, Diag{Line: line, Msg: "`size` requires WxH (e.g. 480x320)"})
				continue
			}
			w, _ := strconv.Atoi(sm[1])
			h, _ := strconv.Atoi(sm[2])
			out.Size = &Size{W: w, H: h}
			continue
		}
		if !tok.q {
			if fm := flagRe.FindStringSubmatch(tok.v); fm != nil {
				val := flagQuoteRe.ReplaceAllString(fm[2], "$1")
				val = strings.ReplaceAll(val, `""`, `"`)
				out.Flags[fm[1]] = val
				continue
			}
		}
		out.Positional = append(out.Positional, tok.v)
	}
	return out
}

var fenceOpenRe = regexp.MustCompile("^(`{3,})\\{([A-Za-z][A-Za-z0-9-]*)\\}(.*)$")

func parseFence(lines []string, i int, m []string, errors *[]Diag) (*FenceBlock, int) {
	open := len(m[1])
	kind := m[2]
	args := ParseInfoArgs(m[3], i+1, errors)
	var body []string
	j := i + 1
	closed := false
	closeRe := regexp.MustCompile(fmt.Sprintf("^`{%d,}\\s*$", open))
	for j < len(lines) {
		if closeRe.MatchString(lines[j]) {
			closed = true
			j++
			break
		}
		body = append(body, lines[j])
		j++
	}
	if !closed {
		*errors = append(*errors, Diag{Line: i + 1, Msg: fmt.Sprintf("unclosed {%s} fence", kind)})
	}
	block := &FenceBlock{Line: i + 1, Kind: kind, Args: args, Body: body}
	refineFence(block, errors)
	return block, j
}

func parseRows(bodyLines []string, baseLine int, errors *[]Diag) []Row {
	var rows []Row
	for k, l := range bodyLines {
		if strings.TrimSpace(l) == "" {
			continue
		}
		cells, ok := SplitPipeRow(l)
		if !ok {
			snippet := l
			if len(snippet) > 50 {
				snippet = snippet[:50]
			}
			*errors = append(*errors, Diag{Line: baseLine + k + 1, Msg: "expected a pipe row, got: " + snippet})
			continue
		}
		rows = append(rows, Row{Cells: cells, Line: baseLine + k + 1})
	}
	return rows
}

func indexOfSeparator(body []string) int {
	for i, l := range body {
		if l == "---" {
			return i
		}
	}
	return -1
}

func refineFence(block *FenceBlock, errors *[]Diag) {
	kind := block.Kind
	body := block.Body
	line := block.Line
	meta := func(arr []string, off int) interface{} {
		return ParseYaml(strings.Join(arr, "\n"), line+off, errors)
	}
	switch {
	case kind == "grid" || kind == "spill-cache":
		block.Rows = parseRows(body, line, errors)
	case kind == "table":
		d := indexOfSeparator(body)
		if d == -1 {
			*errors = append(*errors, Diag{Line: line, Msg: "{table} requires a `---`-separated payload of pipe rows"})
			block.Meta = meta(body, 1)
			block.Rows = nil
		} else {
			block.Meta = meta(body[:d], 1)
			block.Rows = parseRows(body[d+1:], line+d+1, errors)
		}
	case kind == "script":
		d := indexOfSeparator(body)
		if d == -1 {
			block.Meta = map[string]interface{}{}
			block.Code = strings.Join(body, "\n")
		} else {
			block.Meta = meta(body[:d], 1)
			block.Code = strings.Join(body[d+1:], "\n")
		}
	case kind == "raw" || strings.HasPrefix(kind, "x-"):
		block.Payload = strings.Join(body, "\n")
	default:
		block.Meta = meta(body, 1)
	}
}

func parseAt(lines []string, i int, errors *[]Diag) (*AtBlock, int) {
	line := lines[i]
	rest := line[2:]
	sp := strings.IndexByte(rest, ' ')
	targetText := rest
	inline := ""
	if sp != -1 {
		targetText = rest[:sp]
		inline = strings.TrimSpace(rest[sp+1:])
	}

	// Multiline body (Appendix A dedent rule): maximal run of blank-or-2-space
	// lines, trailing blanks excluded.
	j := i + 1
	taken := 0
	lastTake := 0
	var acc []string
	for j < len(lines) {
		l := lines[j]
		if strings.TrimSpace(l) == "" {
			acc = append(acc, "")
			j++
			taken++
			continue
		}
		if strings.HasPrefix(l, "  ") {
			acc = append(acc, l[2:])
			j++
			taken++
			lastTake = taken
			continue
		}
		break
	}
	var bodyLines []string
	if lastTake > 0 {
		bodyLines = acc[:lastTake]
	}
	next := i + 1 + lastTake

	block := &AtBlock{Line: i + 1, TargetText: targetText}
	if bodyLines != nil {
		parsed := ParseYaml(strings.Join(bodyLines, "\n"), i+2, errors)
		if m, ok := parsed.(map[string]interface{}); ok {
			block.Body = m
		} else {
			*errors = append(*errors, Diag{Line: i + 2, Msg: "@ directive body must be a YAML mapping"})
		}
	}
	if inline != "" {
		if strings.HasPrefix(inline, "{") && !strings.HasPrefix(inline, "{=") {
			if props, ok := TryProps(inline); ok {
				block.Props = props
				return block, next
			}
		}
		scalarText, propsText, ok := FindPropsSplit(inline)
		if ok {
			if props, pok := TryProps(propsText); pok {
				block.Props = props
				if scalarText != "" {
					st := scalarText
					block.ScalarText = &st
				}
				return block, next
			}
		}
		st := inline
		block.ScalarText = &st
	}
	return block, next
}

var headingRe = regexp.MustCompile(`^# (.+)$`)
var docCommentRe = regexp.MustCompile(`^#{2,}(\s|$)`)

// Parse parses source into a Document. mode is "strict" (default) or "lenient".
func Parse(source string, mode string) *Document {
	if mode == "" {
		mode = "strict"
	}
	doc := &Document{Frontmatter: map[string]interface{}{}, Mode: mode}
	lines := strings.Split(strings.ReplaceAll(source, "\r\n", "\n"), "\n")

	if len(lines) == 0 || lines[0] != "---" {
		doc.Errors = append(doc.Errors, Diag{Line: 1, Msg: "document must begin with `---` YAML frontmatter"})
		return doc
	}
	fmEnd := -1
	for k := 1; k < len(lines); k++ {
		if lines[k] == "---" {
			fmEnd = k
			break
		}
	}
	if fmEnd == -1 {
		doc.Errors = append(doc.Errors, Diag{Line: 1, Msg: "unterminated frontmatter (missing closing `---`)"})
		return doc
	}
	if fm, ok := ParseYaml(strings.Join(lines[1:fmEnd], "\n"), 2, &doc.Errors).(map[string]interface{}); ok {
		doc.Frontmatter = fm
	}

	i := fmEnd + 1
	var cur *Sheet
	push := func(b Block) {
		if cur != nil {
			cur.Blocks = append(cur.Blocks, b)
		} else {
			doc.WorkbookBlocks = append(doc.WorkbookBlocks, b)
		}
	}

	for i < len(lines) {
		line := lines[i]
		if strings.TrimSpace(line) == "" || strings.HasPrefix(line, ">") || docCommentRe.MatchString(line) {
			i++
			continue
		}
		if m := headingRe.FindStringSubmatch(line); m != nil {
			cur = &Sheet{Name: strings.TrimSpace(m[1]), Line: i + 1}
			doc.Sheets = append(doc.Sheets, cur)
			i++
			continue
		}
		if m := fenceOpenRe.FindStringSubmatch(line); m != nil {
			block, next := parseFence(lines, i, m, &doc.Errors)
			push(block)
			i = next
			continue
		}
		if strings.HasPrefix(line, "@ ") {
			block, next := parseAt(lines, i, &doc.Errors)
			push(block)
			i = next
			continue
		}
		snippet := line
		if len(snippet) > 60 {
			snippet = snippet[:60]
		}
		diag := Diag{Line: i + 1, Msg: "unrecognized line: " + snippet}
		if mode == "strict" {
			doc.Errors = append(doc.Errors, diag)
		} else {
			doc.Warnings = append(doc.Warnings, diag)
		}
		i++
	}
	return doc
}

File: python/src/gridmd/parser.py
"""GridMD document parser (SPEC.md §2-§10, Appendix A).

Produces a block tree; semantic checks live in :mod:`gridmd.validate`.

The embedded YAML is restricted to the GridMD safe subset: no anchors, aliases
or explicit tags, and the YAML-1.2-core scalar schema (only ``true``/``false``
are booleans — ``yes``/``no``/``on``/``off`` and sexagesimal ``12:30`` stay
plain strings; ISO timestamps stay strings, never Python ``datetime``).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

import yaml

RESERVED_KINDS = frozenset(
    [
        "sheet", "grid", "spill-cache", "table", "cf", "validation", "filter",
        "chart", "sparklines", "pivot", "slicer", "image", "shape", "textbox",
        "checkbox", "comments", "outline", "page", "query", "script", "scenario",
        "raw",
    ]
)

_IDENT_KEY_RE = re.compile(r"^(x-)?[a-z][a-z0-9-]*$")
_FENCE_OPEN_RE = re.compile(r"^(`{3,})\{([A-Za-z][A-Za-z0-9-]*)\}(.*)$")


# ---- diagnostics + block tree ----
@dataclass
class Diagnostic:
    line: int
    msg: str


@dataclass
class InfoArgs:
    positional: list[str] = field(default_factory=list)
    flags: dict[str, str] = field(default_factory=dict)
    anchor: str | None = None
    size: dict[str, int] | None = None


@dataclass
class Row:
    cells: list[str]
    line: int


@dataclass
class FenceBlock:
    kind: str
    args: InfoArgs
    body: list[str]
    line: int
    type: str = "fence"
    meta: Any = None
    rows: list[Row] | None = None
    code: str | None = None
    payload: str | None = None


@dataclass
class AtBlock:
    target_text: str
    line: int
    type: str = "at"
    scalar_text: str | None = None
    props: Any = None
    body: Any = None


@dataclass
class SheetBlock:
    name: str
    line: int
    blocks: list = field(default_factory=list)


@dataclass
class ParseStats:
    defs: int = 0
    blocks: int = 0


@dataclass
class ParsedDocument:
    frontmatter: Any = field(default_factory=dict)
    workbook_blocks: list = field(default_factory=list)
    sheets: list[SheetBlock] = field(default_factory=list)
    errors: list[Diagnostic] = field(default_factory=list)
    warnings: list[Diagnostic] = field(default_factory=list)
    mode: str = "strict"
    stats: ParseStats | None = None


# ---- YAML safe-subset loader ----
class _GridmdSafeLoader(yaml.SafeLoader):
    """SafeLoader restricted to the GridMD YAML subset with duplicate-key
    detection."""

    def construct_mapping(self, node: Any, deep: bool = False) -> dict:
        # PyYAML only dispatches here for mapping nodes (the map-tag constructor),
        # so no node-kind guard is needed.
        mapping: dict = {}
        for key_node, value_node in node.value:
            key = self.construct_object(key_node, deep=deep)
            if key in mapping:
                raise yaml.constructor.ConstructorError("while constructing a mapping", node.start_mark, f"found duplicate key {key!r}", key_node.start_mark)
            mapping[key] = self.construct_object(value_node, deep=deep)
        return mapping


# Replace YAML-1.1 implicit resolvers with the YAML-1.2 core scalar schema.
_GridmdSafeLoader.yaml_implicit_resolvers = {}
_GridmdSafeLoader.add_implicit_resolver(
    "tag:yaml.org,2002:bool",
    re.compile(r"^(?:true|True|TRUE|false|False|FALSE)$"),
    list("tTfF"),
)
_GridmdSafeLoader.add_implicit_resolver(
    "tag:yaml.org,2002:int",
    re.compile(r"^(?:[-+]?[0-9]+|0o[0-7]+|0x[0-9a-fA-F]+)$"),
    list("-+0123456789"),
)
_GridmdSafeLoader.add_implicit_resolver(
    "tag:yaml.org,2002:float",
    re.compile(
        r"^(?:[-+]?(?:\.[0-9]+|[0-9]+(?:\.[0-9]*)?)(?:[eE][-+]?[0-9]+)?"
        r"|[-+]?\.(?:inf|Inf|INF)|\.(?:nan|NaN|NAN))$"
    ),
    list("-+0123456789."),
)
_GridmdSafeLoader.add_implicit_resolver(
    "tag:yaml.org,2002:null",
    re.compile(r"^(?:~|null|Null|NULL|)$"),
    ["~", "n", "N", ""],
)


def _reject_unsafe(text: str) -> str | None:
    """Return an error message if the YAML uses anchors, aliases or tags.

    Both an anchor definition and an alias reference carry ``.anchor`` on their
    parse event (the definition always precedes any use in a valid document), so
    a single anchor check rejects both forms.
    """
    for event in yaml.parse(text, Loader=_GridmdSafeLoader):
        if getattr(event, "anchor", None) is not None:
            return "YAML anchors/aliases are outside the GridMD safe subset"
        if getattr(event, "tag", None) is not None:
            return "YAML tags are outside the GridMD safe subset"
    return None


def _load_yaml(text: str) -> Any:
    return yaml.load(text, Loader=_GridmdSafeLoader)


def parse_yaml(text: str, line: int, errors: list[Diagnostic]) -> Any:
    if text.strip() == "":
        return {}
    try:
        unsafe = _reject_unsafe(text)
        if unsafe:
            errors.append(Diagnostic(line, unsafe))
        value = _load_yaml(text)
    except yaml.YAMLError as e:  # malformed YAML
        first = str(e).split("\n")[0]
        errors.append(Diagnostic(line, f"YAML: {first}"))
        return {}
    return {} if value is None else value


def try_props(text: str) -> Any:
    """YAML flow-map candidate for @-directive props: parses to a mapping in
    which every top-level key is an identifier and every value is non-null."""
    try:
        if _reject_unsafe(text) is not None:
            return None
        v = _load_yaml(text)
    except yaml.YAMLError:
        return None
    if v is None or not isinstance(v, dict):
        return None
    for k, val in v.items():
        if not isinstance(k, str) or not _IDENT_KEY_RE.match(k):
            return None
        if val is None:
            return None
    return v


def find_props_split(text: str) -> tuple[str, str | None]:
    """Right-edge props split (SPEC §9.1 / Appendix A): the last brace-balanced
    ``{...}`` group (double quotes respected) that runs to end-of-line and is
    preceded by whitespace."""
    if not text.endswith("}"):
        return text, None
    in_q = False
    depth = 0
    start = -1
    last_group: tuple[int, int] | None = None
    for i, ch in enumerate(text):
        if ch == '"':
            in_q = not in_q
            continue
        if in_q:
            continue
        if ch == "{":
            if depth == 0:
                start = i
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0 and start != -1:
                last_group = (start, i)
            if depth < 0:
                return text, None
    if not last_group:
        return text, None
    s, e = last_group
    if e != len(text) - 1 or s == 0 or text[s - 1] != " ":
        return text, None
    return text[:s].rstrip(), text[s:]


def split_pipe_row(raw_line: str) -> list[str] | None:
    """Pipe row → trimmed cell strings; backslash escapes the next character.
    Returns ``None`` if the line is not a well-formed pipe row."""
    line = re.sub(r"\s+$", "", raw_line)
    if not line.startswith("|") or len(line) < 2:
        return None
    cells: list[str] = []
    cell = ""
    opened = False
    i = 0
    while i < len(line):
        ch = line[i]
        if ch == "\\" and i + 1 < len(line):
            cell += line[i + 1]
            i += 2
            continue
        if ch == "|":
            if not opened:
                opened = True
                i += 1
                continue
            cells.append(cell.strip())
            cell = ""
            i += 1
            continue
        cell += ch
        i += 1
    if cell.strip() != "":
        return None  # no unescaped closing pipe
    return cells


_INFO_TOKEN_RE = re.compile(r'"((?:[^"]|"")*)"|\S+')
_FLAG_RE = re.compile(r"^([A-Za-z][A-Za-z0-9-]*)=(.*)$", re.DOTALL)
_SIZE_RE = re.compile(r"^(\d+)x(\d+)$")


def parse_info_args(rest: str, line: int, errors: list[Diagnostic]) -> InfoArgs:
    """Fence info string: positional args (quoted ``""`` doubling), at-anchors,
    size WxH, key=val flags."""
    out = InfoArgs()
    tokens: list[tuple[str, bool]] = []
    for m in _INFO_TOKEN_RE.finditer(rest):
        if m.group(1) is not None:
            tokens.append((m.group(1).replace('""', '"'), True))
        else:
            tokens.append((m.group(0), False))
    k = 0
    while k < len(tokens):
        value, quoted = tokens[k]
        if not quoted and value == "at":
            k += 1
            if k >= len(tokens):
                errors.append(Diagnostic(line, "`at` requires an anchor"))
                break
            out.anchor = tokens[k][0]
            k += 1
            continue
        if not quoted and value == "size":
            k += 1
            sm = _SIZE_RE.match(tokens[k][0]) if k < len(tokens) else None
            if not sm:
                errors.append(Diagnostic(line, "`size` requires WxH (e.g. 480x320)"))
                k += 1
                continue
            out.size = {"w": int(sm.group(1)), "h": int(sm.group(2))}
            k += 1
            continue
        fm = _FLAG_RE.match(value) if not quoted else None
        if fm:
            flag_val = re.sub(r'^"(.*)"$', r"\1", fm.group(2), flags=re.DOTALL).replace('""', '"')
            out.flags[fm.group(1)] = flag_val
            k += 1
            continue
        out.positional.append(value)
        k += 1
    return out


def _meta(arr: list[str], line: int, off: int, errors: list[Diagnostic]) -> Any:
    return parse_yaml("\n".join(arr), line + off, errors)


def _refine_fence(block: FenceBlock, errors: list[Diagnostic]) -> None:
    kind, body, line = block.kind, block.body, block.line
    if kind in ("grid", "spill-cache"):
        block.rows = _parse_rows(body, line, errors)
    elif kind == "table":
        d = body.index("---") if "---" in body else -1
        if d == -1:
            errors.append(Diagnostic(line, "{table} requires a `---`-separated payload of pipe rows"))
            block.meta = _meta(body, line, 1, errors)
            block.rows = []
        else:
            block.meta = _meta(body[:d], line, 1, errors)
            block.rows = _parse_rows(body[d + 1 :], line + d + 1, errors)
    elif kind == "script":
        d = body.index("---") if "---" in body else -1
        if d == -1:
            block.meta = {}
            block.code = "\n".join(body)
        else:
            block.meta = _meta(body[:d], line, 1, errors)
            block.code = "\n".join(body[d + 1 :])
    elif kind == "raw" or kind.startswith("x-"):
        block.payload = "\n".join(body)
    else:
        block.meta = _meta(body, line, 1, errors)


def _parse_rows(body_lines: list[str], base_line: int, errors: list[Diagnostic]) -> list[Row]:
    rows: list[Row] = []
    for k, line_text in enumerate(body_lines):
        if line_text.strip() == "":
            continue
        cells = split_pipe_row(line_text)
        if cells is None:
            errors.append(Diagnostic(base_line + k + 1, f"expected a pipe row, got: {line_text[:50]}"))
            continue
        rows.append(Row(cells=cells, line=base_line + k + 1))
    return rows


def _parse_fence(lines: list[str], i: int, m: re.Match, errors: list[Diagnostic]) -> tuple[FenceBlock, int]:
    open_len = len(m.group(1))
    kind = m.group(2)
    args = parse_info_args(m.group(3) or "", i + 1, errors)
    body: list[str] = []
    j = i + 1
    closed = False
    close_re = re.compile(rf"^`{{{open_len},}}\s*$")
    while j < len(lines):
        if close_re.match(lines[j]):
            closed = True
            j += 1
            break
        body.append(lines[j])
        j += 1
    if not closed:
        errors.append(Diagnostic(i + 1, f"unclosed {{{kind}}} fence"))
    block = FenceBlock(kind=kind, args=args, body=body, line=i + 1)
    _refine_fence(block, errors)
    return block, j


def _parse_at(lines: list[str], i: int, errors: list[Diagnostic]) -> tuple[AtBlock, int]:
    line = lines[i]
    rest = line[2:]
    sp = rest.find(" ")
    target_text = rest if sp == -1 else rest[:sp]
    inline = "" if sp == -1 else rest[sp + 1 :].strip()

    # Multiline body: maximal run of blank-or-2-space-indented lines, trailing
    # blanks excluded (Appendix A, dedent rule).
    j = i + 1
    taken = 0
    last_take = 0
    acc: list[str] = []
    while j < len(lines):
        current = lines[j]
        if current.strip() == "":
            acc.append("")
            j += 1
            taken += 1
            continue
        if re.match(r"^ {2}", current):
            acc.append(current[2:])
            j += 1
            taken += 1
            last_take = taken
            continue
        break
    body_lines = acc[:last_take] if last_take > 0 else None
    nxt = i + 1 + last_take

    block = AtBlock(target_text=target_text, line=i + 1)
    if body_lines is not None:
        parsed = parse_yaml("\n".join(body_lines), i + 2, errors)
        if parsed is None or not isinstance(parsed, dict):
            errors.append(Diagnostic(i + 2, "@ directive body must be a YAML mapping"))
        else:
            block.body = parsed
    if inline != "":
        if inline.startswith("{") and not inline.startswith("{="):
            props = try_props(inline)
            if props:
                block.props = props
                return block, nxt
        scalar_text, props_text = find_props_split(inline)
        if props_text:
            props = try_props(props_text)
            if props:
                block.props = props
                block.scalar_text = None if scalar_text == "" else scalar_text
                return block, nxt
        block.scalar_text = inline
    return block, nxt


def parse_document(source: str, mode: str = "strict") -> ParsedDocument:
    errors: list[Diagnostic] = []
    warnings: list[Diagnostic] = []
    lines = re.split(r"\r?\n", source)
    doc = ParsedDocument(errors=errors, warnings=warnings, mode=mode)

    if not lines or lines[0] != "---":
        errors.append(Diagnostic(1, "document must begin with `---` YAML frontmatter"))
        return doc
    fm_end = -1
    for k in range(1, len(lines)):
        if lines[k] == "---":
            fm_end = k
            break
    if fm_end == -1:
        errors.append(Diagnostic(1, "unterminated frontmatter (missing closing `---`)"))
        return doc
    doc.frontmatter = parse_yaml("\n".join(lines[1:fm_end]), 2, errors)

    i = fm_end + 1
    cur: SheetBlock | None = None

    def push(b: Any) -> None:
        (cur.blocks if cur else doc.workbook_blocks).append(b)

    while i < len(lines):
        line = lines[i]
        if line.strip() == "" or re.match(r"^>", line) or re.match(r"^#{2,}(\s|$)", line):
            i += 1
            continue
        m = re.match(r"^# (.+)$", line)
        if m:
            cur = SheetBlock(name=m.group(1).strip(), line=i + 1)
            doc.sheets.append(cur)
            i += 1
            continue
        m = _FENCE_OPEN_RE.match(line)
        if m:
            block, nxt = _parse_fence(lines, i, m, errors)
            push(block)
            i = nxt
            continue
        if re.match(r"^@ ", line):
            block, nxt = _parse_at(lines, i, errors)
            push(block)
            i = nxt
            continue
        target = errors if mode == "strict" else warnings
        target.append(Diagnostic(i + 1, f"unrecognized line: {line[:60]}"))
        i += 1
    return doc
